### PR TITLE
Switch to Debian 11 (bullseye) as base for our dockerfiles

### LINKED ISSUE #18190: Docker image - Migrate to 3.x-slim-bullseye from 3.x-slim-buster
### Description

Currently docker images are produced based on top of 3.x-slim-buster python and as a result output inherits related debian vulnerabilities. 

### Use case/motivation

Migrate to the latest debian release (bullseye was released around a month ago) for docker images.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by potiuk**
Debian 11 Bullseye have been released some time ago as the new
LTS Debian release and already all our dependencies (including
MySQL and MSSQL ODBC drivers) caught up with it so we can finally
migrate to it.

This change switches base images to bullsey for our Dockerfiles
as well as for Redis image we are using in CI.

The relevant packages have been updated to include that
and documentation have been updated.

Examples of ours also are updated to use "bullseye" rather than
buster.

Closes: #18190
Closes: #18279

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: Dockerfile.ci [hunk 1]
```diff
@@ -78,7 +78,6 @@ RUN apt-get update \
            libsasl2-dev \
            libsasl2-modules \
            libssl-dev \
-           libenchant-dev \
```
Comment (potiuk) [inline Dockerfile.ci]: I believe this one is not needed any more. Let's see if all tests pass (this is a spellcheck library with some rather obscure use).

**Comment by potiuk** [comment]
This one will not succeed, because the "main" scripts use "buster" as a base and override the change in this PR (but the images are building fine locally when I run them): 

To test it - I am running it in "main" of my fork:

* Build workflow: https://github.com/potiuk/airflow/actions/runs/1805858131
* Test workflow: https://github.com/potiuk/airflow/actions/runs/1805858128

**Comment by potiuk** [comment]
Few more changes are needed it seems :)

File: Dockerfile.ci (hunk 1 above)
Comment (potiuk) [reply Dockerfile.ci]: Ah no... It's used by sphinx :(

**Comment by potiuk** [comment]
Looks like (at least from the first trial) debian bullseye might be a bit more resource (memory?) hungry and a lot of tests gets less stable in the Public Runners. I will need to investigate that.

**Comment by potiuk** [comment]
Unfortuntely we cannot move to bullseye (yet). The problem is that MSSQL ODBC driver (which is the only "reliable" and stable drirver for MSSQL for SQLAlchemy) does not support Debian Bullseye yet.

This has been documented and commented on by MSSQL driver maintainers:

- https://github.com/MicrosoftDocs/sql-docs/issues/6494
- https://github.com/MicrosoftDocs/sql-docs/issues/6789
- https://github.com/MicrosoftDocs/sql-docs/issues/6804
- https://github.com/MicrosoftDocs/sql-docs/issues/7255

The answer in all those issues given 28 Jan 2022:

> Debian 11 installation instructions will be added when msodbcsql17 supports. Debian 11.
> Regards,
> David

Let's hope it will be soon. This is the last blocker from migrating to Bullseye. All the rest seems to work

I asked the maintainers for the expected timeline https://github.com/MicrosoftDocs/sql-docs/issues/7255#issuecomment-1037097131

**Comment by potiuk** [comment]
Now - the actual switch to Bullseye is here :)

**Comment by potiuk** [comment]
I wil also add later a separate PR where the Buster images will also be built (but I want to do with the new Breeze2 python build image to not complicate the bash scripts any more).

**Review by ashb — APPROVED** [review]

**Comment by potiuk** [comment]
Just random I/O timeouts merging.

**Pull request merged**